sql/src/main/java/io/crate/action/sql/Session.java

```diff
@@ -391,49 +388,209 @@ public void execute(String portalName, int maxRows, ResultReceiver resultReceive
                 }
                 preparedStatements.clear();
             }
-            portal.sync(planner, jobsLogs);
+            resultReceiver.allFinished(false);
         } else {
-            // delay execution to be able to bundle bulk operations
-            pendingExecutions.add(portal);
+            /* We defer the execution for any other statements to `sync` messages so that we can efficiently process
+             * bulk operations. E.g. If we receive `INSERT INTO (x) VALUES (?)` bindings/execute multiple times
+             * We want to create bulk requests internally:                                                          /
+             * - To reduce network overhead
+             * - To have 1 disk flush per shard instead of 1 disk flush per item
+             *
+             * Many clients support this by doing something like this:
+             *
+             *      var preparedStatement = conn.prepareStatement("...")
+             *      for (var args in manyArgs):
+             *          preparedStatement.execute(args)
+             *      conn.commit()
+             */
+            deferredExecutions.add(new DeferredExecution(portal, maxRows, resultReceiver));
         }
     }
 
     public CompletableFuture<?> sync() {
-        switch (pendingExecutions.size()) {
+        switch (deferredExecutions.size()) {
             case 0:
                 LOGGER.debug("method=sync pendingExecutions=0");
                 return CompletableFuture.completedFuture(null);
+
             case 1:
-                Portal portal = pendingExecutions.iterator().next();
+                DeferredExecution deferredExecution = deferredExecutions.get(0);
+                deferredExecutions.clear();
+                Portal portal = deferredExecution.portal();
                 LOGGER.debug("method=sync portal={}", portal);
-                pendingExecutions.clear();
-                clearState();
-                return portal.sync(planner, jobsLogs);
+                return singleExec(portal, deferredExecution.resultReceiver(), deferredExecution.maxRows());
+
             default:
-                throw new IllegalStateException(
-                    "Shouldn't have more than 1 pending execution. Got: " + pendingExecutions);
+                List<DeferredExecution> deferredExecutions = List.copyOf(this.deferredExecutions);
+                this.deferredExecutions.clear();
+                Map<Statement, List<DeferredExecution>> deferredExecutionsByStatement = deferredExecutions
```

Comment:
We could eagerly create the grouped structure and probably also mix singleExec and bulkExec within the same `sync` call. 

But I'd keep it for now like this for the backport and followup on this on master only.